### ISSUE TITLE
Spark core concepts explained

### ISSUE THREAD

**Issue by guevara** [guevara/read-it-later#4615]
Spark core concepts explained<br>
<br>
https://ift.tt/2xeuG7K<br>
<br>
About 0 Minutes<br>
<br>
<div>
<div>
<img src="https://luminousmen.com/media/spark.jpg" alt="Spark core concepts explained">
<p><a href="https://spark.apache.org/">Apache Spark</a>&nbsp;is considered as a powerful complement to&nbsp;<a href="https://hadoop.apache.org/">Hadoop</a>, big data&rsquo;s original technology. Spark is a more accessible, powerful and capable big data tool for tackling various big data challenges. It has become mainstream and most in-demand big data framework across all major industries. Spark has become part of the Hadoop since 2.0. And is one of the most useful technologies for Python Big Data Engineers.</p>
<p>This series of posts is a single-stop resource that gives spark architecture overview and it's good for people looking to learn spark.</p>
<p>Whole series:</p>
<hr>
<p>Apache Spark Architecture is based on two main abstractions:</p>
<ul>
<li>Resilient Distributed Dataset (RDD)</li>
<li>Directed Acyclic Graph (DAG)</li>
</ul>
<p>Let's dive in these concepts</p>
<h2>RDD &mdash; the Spark basic concept</h2>
<p>The key to understanding Apache Spark is RDD &mdash; the Resilient Distributed Dataset. RDD contains an <strong>arbitrary collection of objects</strong>. Each data set in RDD is logically distributed across the cluster nodes so that they can be processed in parallel.</p>
<p>Physically, RDD is stored as an object in driver JVM and it refers to the data stored either in persisted store (HDFS, Cassandra, HBase, etc.) or in the cache (memory, memory+disks, disk only, etc.) or in another RDD.</p>
<p>RDD stores the following metadata:
 &#9642; Partitions &ndash; set of data splits associated with this RDD. They are located on the cluster nodes. One partition is a minimal data batch which will be processed by each cluster node;
 &#9642; Dependencies &ndash; list of parent RDDs involved in the computation aka lineage graph;
 &#9642; Computation &ndash; function to compute child RDD given the parent RDD from the Dependencies;
 &#9642; Preferred Locations &ndash; where is the best place to put computations on partitions (data locality);
 &#9642; Partitioner &ndash; how the data is split into Partitions(by default they split by <code>HashPartitioner</code>);</p>
<p>RDD can be recreated as well as data that it refers to because every RDD knows how it was created (by storing the lineage graph). RDD also can be materialized, in memory or on disk.</p>
<p>Example:</p>
<pre>
<code> &gt;&gt;&gt; rdd = sc.parallelize(range(20)) # create RDD
 &gt;&gt;&gt; rdd
 ParallelCollectionRDD[0] at parallelize at PythonRDD.scala:195
 &gt;&gt;&gt; rdd.collect() # collect data on driver and show
 [0, 1, 2, 3, 4, 5, 6, 7, 8, 9, 10, 11, 12, 13, 14, 15, 16, 17, 18, 19]
 </code>
</pre>
<p>RDD can also be cached and manually partitioned. Caching is beneficial when we use RDD several times(and may slow down our computations otherwise). And manual partitioning is important to correctly balance data between partitions. Generally, smaller partitions allow distributing data more equally, among more executors. Hence, fewer partitions may boost tasks with a lot of repartitions(data reorganizations during computations).</p>
<p>Let's check the number of partitions and data on them:</p>
<pre>
<code> &gt;&gt;&gt; rdd.getNumPartitions() # get current number of paritions
 4
 &gt;&gt;&gt; rdd.glom().collect() # collect data on driver based on partitions
 [[0, 1, 2, 3, 4], [5, 6, 7, 8, 9], [10, 11, 12, 13, 14], [15, 16, 17, 18, 19]]
 </code>
</pre>
<p>All interesting that happens in Spark happens through RDD operations. That is, usually Spark applications look like the following &mdash; we create RDD (for example, we set data source as file on HDFS), we transform it (<code>map</code>, <code>reduce</code>, <code>join</code>, <code>groupBy</code>, <code>aggregate</code>, <code>reduce</code>, ...), do something with the result (for example, we throw it back into HDFS).</p>
<p><img src="https://luminousmen.com/media/spark_application_flow.jpg" alt="Spark Application Flow"></p>
<p>Over RDD, you can do two types of operations (and, accordingly, all the work with the data is in the sequence of these two types): <strong>transformations and actions</strong>.</p>
<h3><strong>Transformations</strong></h3>
<p>The result of applying this operation to RDD is a <strong>new RDD</strong>. As a rule, these are operations that in some way convert the elements of a given data.</p>
<p>Transformations are <strong>lazy in nature</strong> meaning when we call some operation in RDD, it does not execute immediately. Spark maintains the record of which operation is being called(through DAG, we will talk about it later).</p>
<p>We can think Spark RDD as the data, that we built up through transformation. Because of transformations laziness, we can execute operation any time by calling an action on data. Hence, data is not loaded until it is necessary. It gives plenty of opportunities to induce low-level optimizations.</p>
<p>At a high level, there are two groups of transformations that can be applied onto the RDDs, namely <strong>narrow transformations</strong>, and <strong>wide transformations</strong>.</p>
<p><img src="https://luminousmen.com/media/narrow_wide_transformations.jpg" alt="Narrow/wide transformations"></p>
<p><strong>Narrow transformation</strong> doesn't require the data to be <strong>shuffled or reorganized</strong> across the partitions. For example, <code>map</code>, <code>filter</code>, etc. The narrow transformations will be grouped (or pipe-lined) together into a single stage.</p>
<p>A&nbsp;<strong>shuffle</strong>&nbsp;occurs when data is rearranged between partitions. This is required when a transformation requires information from other partitions, such as summing all the values in a column. Spark will gather the required data from each partition and combine it into a new partition, likely on a different executor.</p>
<p>But there are exceptions, operations like <code>coalesce</code> may cause the task to work with multiple input partitions, but the transformation will still be considered narrow because the input records used to compute any output record can still be found only in a limited subset of partitions.</p>
<p>Let's use filter transformation on our data:</p>
<pre>
<code> &gt;&gt;&gt; filteredRDD = rdd.filter(lambda x: x &gt; 10)
 &gt;&gt;&gt; print(filteredRDD.toDebugString()) # to see the execution graph; only one stage is created
 (4) PythonRDD[1] at RDD at PythonRDD.scala:53 []
 | ParallelCollectionRDD[0] at parallelize at PythonRDD.scala:195 []
 &gt;&gt;&gt; filteredRDD.collect()
 [11, 12, 13, 14, 15, 16, 17, 18, 19]
 </code>
</pre>
<p>In this example we don't need to shuffle the data, each partition can be processed independently.</p>
<p>However, Spark also supports transformations with wide dependencies(namely <strong>wide transformations</strong>), such as <code>groupByKey</code>, <code>reduceByKey</code>, etc. Within such dependencies, the data required for calculation may be located in several partitions of the parent RDD. All data with the same key must be in the same partition, processed by a single task. To implement these operations, Spark must perform the shuffling, moving data across the cluster and forming a new stage with a new set of partitions as in the example below:</p>
<pre>
<code> &gt;&gt;&gt; groupedRDD = filteredRDD.groupBy(lambda x: x % 2) # group data based on mod
 &gt;&gt;&gt; print(groupedRDD.toDebugString()) # two separate stages are created, because of the shuffle
 (4) PythonRDD[6] at RDD at PythonRDD.scala:53 []
 | MapPartitionsRDD[5] at mapPartitions at PythonRDD.scala:133 []
 | ShuffledRDD[4] at partitionBy at NativeMethodAccessorImpl.java:0 []
 +-(4) PairwiseRDD[3] at groupBy at &lt;ipython-input-5-a92aa13dcb83&gt;:1 []
 | PythonRDD[2] at groupBy at &lt;ipython-input-5-a92aa13dcb83&gt;:1 []
 | ParallelCollectionRDD[0] at parallelize at PythonRDD.scala:195 []
 </code>
</pre>
<h3>Actions</h3>
<p>Actions are applied when it is necessary to materialize the result &mdash; save the data to disk, or output part of the data to the console. <code>collect</code> operation we used so far is also an action &mdash; it collects data.</p>
<p>Actions are not lazy &mdash; they actually will trigger the data processing. Actions are RDD operations that produce non-RDD values.</p>
<p>For example, to get sum of our filtered data we can use <code>reduce</code>:</p>
<pre>
<code> &gt;&gt; filteredRDD.reduce(lambda a, b: a + b)
 135
 </code>
</pre>
<h2>DAG</h2>
<p>Unlike Hadoop where the user has to break down the whole job into smaller jobs and chain them together to go along with MapReduce, Spark identifies the tasks that can be computed in parallel with partitioned data on the cluster. With these identified tasks, Spark builds a logical flow of operations that can be represented in a graph which is directed and acyclic, also known as <strong>DAG</strong> (Directed Acyclic Graph), where a node is RDD partition and the edge is transformation on the data. Thus Spark builds its own plan of executions implicitly from the provided spark application.</p>
<p><img src="https://luminousmen.com/media/dag_flow.jpg" alt="DAG flow"></p>
<p>The <strong>DAGScheduler</strong> divides operators into stages of tasks. A stage is consist of tasks based on the input data partitions. The DAGScheduler pipelines some transformations together. For e.g. many map operators can be squash into a single stage. The final result of a DAGScheduler is a set of stages. The stages are passed on to the <strong>TaskScheduler</strong>. The number of tasks submitted depends on the number of partitions. The TaskScheduler launches tasks via the <strong><a href="https://luminousmen.com/post/hadoop-yarn-spark">cluster manager</a></strong>. The TaskScheduler doesn't know about dependencies of the stages.</p>
<p>RDDs are capable of defining location preference to compute partitions. Location preference refers to information about the RDD location. The&nbsp;<strong>DAGScheduler</strong>&nbsp;places the partitions in such a way that task is close to data as much as possible(data locality).</p>
<a href="https://www.buymeacoffee.com/OheL3Mfnp" target="blank">Support author</a>
</div>
</div>
<br>
<br>
<br>
<br>
<br>
<br>
via Blog | iamluminousmen https://luminousmen.com<br>
<br>
June 28, 2019 at 11:45AM
